File: yoga-imessager/src/main/java/com/yoga/imessager/user/controller/UserApiController.java
package com.yoga.imessager.user.controller;

import com.yoga.core.annotation.Explain;
import com.yoga.core.controller.BaseApiController;
import com.yoga.core.data.CommonResult;
import com.yoga.core.data.PageList;
import com.yoga.core.data.ResultConstants;
import com.yoga.core.exception.IllegalArgumentException;
import com.yoga.core.utils.MapConverter;
import com.yoga.core.utils.StrUtil;
import com.yoga.imessager.group.dto.FindDto;
import com.yoga.imessager.group.model.Group;
import com.yoga.imessager.group.service.GroupService;
import com.yoga.imessager.rongcloud.service.RongService;
import com.yoga.imessager.user.dto.*;
import com.yoga.imessager.user.model.User;
import com.yoga.imessager.user.service.SystemUserService;
import com.yoga.imessager.user.service.UserService;
import com.yoga.user.basic.TenantDto;
import com.yoga.user.basic.TenantPage;
import com.yoga.user.dept.model.Department;
import com.yoga.user.dept.service.DepartmentService;
import com.yoga.user.model.LoginUser;
import org.apache.shiro.SecurityUtils;
import org.apache.shiro.authz.annotation.RequiresAuthentication;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;
import java.util.List;

@Explain(value = "IM用户管理", module = "gcf_imessager")
@RestController(value = "imUserApiController")
@RequestMapping("/api/im/user")
public class UserApiController extends BaseApiController{

    @Autowired
    private RongService rongService;
    @Autowired
    private UserService userService;
    @Autowired
    private GroupService groupService;
    @Autowired
    private SystemUserService systemUserService;
    @Autowired
    private DepartmentService departmentService;

    @Explain("获取Token")
    @RequestMapping("/token")
    public CommonResult getToken(TenantDto dto) throws Exception {
        if (!SecurityUtils.getSubject().isAuthenticated() || SecurityUtils.getSubject().getSession() == null) new CommonResult(ResultConstants.ERROR_UNAUTHENTICAED, "请登录");
        LoginUser user = new LoginUser();
        String name = StrUtil.isBlank(user.getNickname()) ? user.getUsername() : user.getNickname();
        String token = rongService.token(dto.getTid(), String.valueOf(user.getId()), name, user.getAvatar());
        return new CommonResult(token);
    }

    @Explain("获取个人信息")
    @RequiresAuthentication
    @RequestMapping("/info")
    public CommonResult getInfo(HttpServletRequest request, LoginUser loginUser, @Valid InfoDto dto, BindingResult bindingResult) throws Exception {
        if (bindingResult.hasErrors()) throw new IllegalArgumentException(bindingResult);
        long userId = dto.getId() == null ? loginUser.getId() : dto.getId();
        User user = userService.get(dto.getTid(), userId);
        return new CommonResult(new MapConverter<>(new MapConverter.Converter<User>() {
            @Override
            public void convert(User item, MapConverter.MapItem<String, Object> map) {
                map.set("id", item.getId())
                        .set("nickname", item.getNickname())
                        .set("avatar", item.getAvatar());
            }
        }).build(user));
    }

    @Explain("更新个人信息")
    @RequiresAuthentication
    @RequestMapping("/update")
    public CommonResult updateInfo(HttpServletRequest request, LoginUser loginUser, @Valid UpdateDto dto, BindingResult bindingResult) throws Exception {
        if (bindingResult.hasErrors()) throw new IllegalArgumentException(bindingResult);
        userService.update(dto.getTid(), loginUser.getId(), dto.getNickname(), dto.getAvatar());
        return new CommonResult();
    }


    @Explain("查找用户信息")
    @RequiresAuthentication
    @RequestMapping("/find")
    public CommonResult findUser(HttpServletRequest request, TenantPage page, @Valid FindDto dto, BindingResult bindingResult) throws Exception {
        if (bindingResult.hasErrors()) throw new IllegalArgumentException(bindingResult);
        Page<User> users = userService.find(dto.getTid(), dto.getName(), page.getPageIndex(), page.getPageSize());
        return new CommonResult(new MapConverter<>(new MapConverter.Converter<User>() {
            @Override
            public void convert(User item, MapConverter.MapItem<String, Object> map) {
                map.set("id", item.getId())
                        .set("nickname", item.getNickname())
                        .set("avatar", item.getAvatar());
            }
        }).build(users), users);
    }

    @Explain("查找系统用户")
    @RequiresAuthentication
    @RequestMapping("/find/system")
    public CommonResult findSystemUser(HttpServletRequest request, TenantPage page, @Valid FindSystemDto dto, BindingResult bindingResult) throws Exception {
        if (bindingResult.hasErrors()) throw new IllegalArgumentException(bindingResult);
        PageList<com.yoga.user.user.model.User> users = systemUserService.findUsers(dto.getTid(), dto.getGroupId(), dto.getNickname(), dto.getDeptId(), page.getPageIndex(), page.getPageSize());
        return new CommonResult(new MapConverter<>(new MapConverter.Converter<com.yoga.user.user.model.User>() {
            @Override
            public void convert(com.yoga.user.user.model.User item, MapConverter.MapItem<String, Object> map) {
                map.set("id", item.getId())
                        .set("username", item.getUsername())
                        .set("firstname", item.getFirstname())
                        .set("lastname", item.getLastname())
                        .set("fullname", item.getFullname())
                        .set("email", item.getEmail())
                        .set("qq", item.getQq())
                        .set("mobile", item.getPhone())
                        .set("avatar", item.getAvatar())
                        .set("title", item.getTitle());
                Department department = departmentService.get(dto.getTid(), item.getDeptId());
                if (department != null) {
                    map.set("departmentId", department.getId())
                            .set("department", department.getName());
                }
            }
        }).build(users), users.getPage());
    }

    @Explain("获取群组列表")
    @RequiresAuthentication
    @RequestMapping("/groups")
    public CommonResult getGoups(HttpServletRequest request, LoginUser loginUser, @Valid TenantDto dto, BindingResult bindingResult) throws Exception{
        if (bindingResult.hasErrors()) throw new IllegalArgumentException(bindingResult);
        if (groupService.everyOneGroup(dto.getTid())) {
            groupService.ensureUserInEveryOneGroup(dto.getTid(), loginUser);
        }
        List<Group> groups = userService.groups(dto.getTid(), loginUser.getId());
        return new CommonResult(new MapConverter<>(new MapConverter.Converter<Group>() {
            @Override
            public void convert(Group item, MapConverter.MapItem<String, Object> map) {
                map.set("id", item.getId())
                        .set("name", item.getName())
                        .set("avatar", item.getAvatar())
                        .set("memberCount", item.getMemberCount())
                        .set("createTime", item.getCreateTime())
                        .set("creator", item.getCreator())
                        .set("creatorId", item.getCreatorId());
            }
        }).build(groups));
    }

    @Explain("申请加入群组")
    @RequiresAuthentication
    @RequestMapping("/apply")
    public CommonResult applyJoin(HttpServletRequest request, LoginUser login, @Valid ApplyDto dto, BindingResult bindingResult) throws Exception {
        if (bindingResult.hasErrors()) throw new IllegalArgumentException(bindingResult);
        groupService.apply(dto.getTid(), dto.getId(), login.getId(), login.getNickname(), login.getAvatar(), dto.getMessage());
        return new CommonResult();
    }

    @Explain("退出群组")
    @RequiresAuthentication
    @RequestMapping("/quit")
    public CommonResult quitGroup(HttpServletRequest request, LoginUser loginUser, @Valid QuitDto dto, BindingResult bindingResult) throws Exception {
        if (bindingResult.hasErrors()) throw new IllegalArgumentException(bindingResult);
        groupService.quit(dto.getTid(), dto.getId(), loginUser.getId());
        return new CommonResult();
    }
}

File: yoga-wechat/src/main/kotlin/com/yoga/wechat/users/UserApiController.kt
package com.yoga.wechat.users

import com.yoga.core.annotation.Explain
import com.yoga.core.data.CommonResult
import com.yoga.core.exception.IllegalArgumentException
import com.yoga.user.model.LoginUser
import com.yoga.wechat.base.WxBaseApiController
import org.apache.shiro.authz.annotation.RequiresAuthentication
import org.apache.shiro.authz.annotation.RequiresPermissions
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.validation.BindingResult
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import javax.validation.Valid

@Explain("WX用户管理", module = "wx_users")
@RestController(value = "wechatUserApiController")
@RequestMapping("/api/wechat/users")
open class UserApiController @Autowired constructor(
        val userService: UserService) : WxBaseApiController() {

    @Explain("从公众号刷新")
    @RequiresPermissions("wx_users.update")
    @RequestMapping("/refresh")
    open fun refresh(user: LoginUser, @Valid dto: RefreshDto, bindingResult: BindingResult): CommonResult {
        if (bindingResult.hasErrors()) throw IllegalArgumentException(bindingResult)
        userService.sync(dto.tid, dto.accountId!!, user.nickname, user.id)
        return CommonResult()
    }

    @Explain("修改用户备注")
    @RequiresPermissions("wx_users.update")
    @RequestMapping("/update")
    open fun update(@Valid dto: UpdateDto, bindingResult: BindingResult): CommonResult {
        if (bindingResult.hasErrors()) throw IllegalArgumentException(bindingResult)
        userService.update(dto.tid, dto.accountId!!, dto.openId!!, dto.remark)
        return CommonResult()
    }

    @Explain("绑定用户信息")
    @RequiresAuthentication
    @RequestMapping("/bind")
    open fun binding(loginUser: LoginUser, @Valid dto: BindingDto, bindingResult: BindingResult): CommonResult {
        if (bindingResult.hasErrors()) throw IllegalArgumentException(bindingResult)
        userService.bind(dto.tid, loginUser.id, dto.accountId!!, dto.openId!!)
        return CommonResult()
    }

    @Explain("用户TAG列表")
    @RequiresPermissions("wx_users.update")
    @RequestMapping("/tags")
    open fun tags(@Valid dto: TagsDto, bindingResult: BindingResult): CommonResult {
        if (bindingResult.hasErrors()) throw IllegalArgumentException(bindingResult)
        val tags = userService.tags(dto.tid, dto.accountId!!, dto.openId!!)
        return CommonResult(tags)
    }
    @Explain("删除用户标签")
    @RequiresPermissions("wx_users.update")
    @RequestMapping("/untag")
    open fun untag(@Valid dto: UntagDto, bindingResult: BindingResult): CommonResult {
        if (bindingResult.hasErrors()) throw IllegalArgumentException(bindingResult)
        userService.removeTag(dto.tid, dto.accountId!!, dto.tagId!!, arrayOf(dto.openId!!))
        return CommonResult()
    }
    @Explain("添加用户标签")
    @RequiresPermissions("wx_users.update")
    @RequestMapping("/tag")
    open fun tag(@Valid dto: TagDto, bindingResult: BindingResult): CommonResult {
        if (bindingResult.hasErrors()) throw IllegalArgumentException(bindingResult)
        userService.addTag(dto.tid, dto.accountId!!, dto.tagId!!, arrayOf(dto.openId!!))
        return CommonResult()
    }
}
